pyro/ops/gaussian.py

```diff
@@ -155,25 +155,37 @@ def condition(self, value):
                           b.mul(info_b).sum(-1))
         return Gaussian(log_normalizer, info_vec, precision)
 
-    def marginalize(self, slice_):
+    def marginalize(self, left=0, right=0):
         """
-        Slice along event dimension, marginalizing out dropped variables.
-        This should satisfy for a slice ``s``::
+        Marginalizing out variables on either side of the event dimension::
 
-            g.marginalize(s).event_logsumexp() = g.logsumexp()
+            g.marginalize(left=n).event_logsumexp() = g.logsumexp()
+            g.marginalize(right=n).event_logsumexp() = g.logsumexp()
 
         and for data ``x``:
 
             g.condition(x).event_logsumexp()
-              = g.marginalize(slice(0, x.dim()).log_density(x)
+              = g.marginalize(left=g.dim() - x.size(-1)).log_density(x)
         """
-        assert isinstance(slice_, slice), "advanced indexing is not supported"
-        info_vec = self.info_vec[..., slice_]  # FIXME is this right?
-        precision = self.precision[..., slice_, slice_]
-        n = info_vec.size(-1)
+        if left == 0 and right == 0:
+            return self
+        if left > 0 and right > 0:
+            raise NotImplementedError
+        n = self.dim()
+        n_b = left + right
+        a = slice(left, n - right)  # preserved
+        b = slice(None, left) if left else slice(n - right, None)
+        info_vec = self.info_vec[..., a]  # FIXME
```

Comment:
I think
```
P_at = P_a.transpose(-1, -2)
b_tmp = b.unsqueeze(-1).triangular_solve(P_b, upper=False).solution
info_vec = self.info_vec[..., a] - torch.matmul(P_at, b_tmp).squeeze(-1)
```

Reply:
Yes, I implemented this and re-derived your math to verify. However tests still fail 😕 
- Tests pass if we ignore `info_vec` computations using `gaussian.info_vec.fill_(0)` in  the test examples. This suggests that the `precision` and `log_normalier` math is correct.
- Your snippet provides equivalent results to the brute force
  ```
  info_vec = precision.matmul(
      self.precision.inverse().matmul(self.info_vec.unsqueeze(-1))[..., a, :]).squeeze(-1)
  ```
  This suggests that your snippet is correct.

Any ideas what could be wrong?

Reply:
Just recognize that you'll need to add `0.5 * b_tmp.squeeze(-1).pow(2).sum(-1)` term to `log_normalizer` too. When `info_vec=0`, this term is zero, so you didn't detect this. :D